File: lib/engines/tone_generator.dart
import 'dart:math';
import 'dart:typed_data';

class ToneGenerator {
  static const int _sampleRate = 44100;
  static const int _bitsPerSample = 16;

  static Uint8List tickWav() {
    return _generateWav(frequency: 1800, durationMs: 40, volume: 0.7);
  }

  static Uint8List accentWav() {
    return _generateWav(frequency: 900, durationMs: 60, volume: 0.9);
  }

  // Count-in beeps (different pitch per number)
  static Uint8List countIn3Wav() {
    return _generateWav(frequency: 440, durationMs: 120, volume: 0.7);
  }

  static Uint8List countIn2Wav() {
    return _generateWav(frequency: 660, durationMs: 120, volume: 0.7);
  }

  static Uint8List countIn1Wav() {
    return _generateWav(frequency: 880, durationMs: 120, volume: 0.7);
  }

  static Uint8List countInStartWav() {
    return _generateWav(frequency: 1200, durationMs: 200, volume: 0.9);
  }

  // 10-second countdown warning — three quick beeps
  static Uint8List countdownWarningWav() {
    return _multiBeepWav(1000, 60, 50, 3, 0.8);
  }

  // Final 5-second warning — faster beeps
  static Uint8List countdownFinalWav() {
    return _multiBeepWav(1200, 40, 30, 5, 0.9);
  }

  // End bell — descending chime
  static Uint8List endBellWav() {
    return _chimeWav([1200, 1000, 800, 600], 200, 40, 0.9);
  }

  /// Generic alert sweep (for tone mode).
  static Uint8List signalWav(DirectionSignalType type) {
    switch (type) {
      case DirectionSignalType.alert:
        return _generateSweepWav(600, 1200, 150, 0.9);
      case DirectionSignalType.whistle:
        return _generateSweepWav(1200, 2000, 120, 0.85);
      case DirectionSignalType.drum:
        return _generateWav(frequency: 120, durationMs: 100, volume: 1.0);
      case DirectionSignalType.dirForward:
        return _dirForwardWav();
      case DirectionSignalType.dirBackward:
        return _dirBackwardWav();
      case DirectionSignalType.dirLeft:
        return _dirLeftWav();
      case DirectionSignalType.dirRight:
        return _dirRightWav();
    }
  }

  /// Rapid ascending two-note
  static Uint8List _dirForwardWav() {
    return _twoNoteWav(800, 1300, 45, 55, 0.95);
  }

  /// Rapid descending two-note
  static Uint8List _dirBackwardWav() {
    return _twoNoteWav(1200, 600, 45, 55, 0.95);
  }

  /// Sharp rising chirp
  static Uint8List _dirLeftWav() {
    return _generateSweepWav(1600, 2600, 75, 0.9);
  }

  /// Sharp falling chirp
  static Uint8List _dirRightWav() {
    return _generateSweepWav(2600, 1400, 75, 0.9);
  }

  /// Multiple quick beeps with gaps
  static Uint8List _multiBeepWav(double freq, int durMs, int gapMs, int count, double volume) {
    final beepSamples = (_sampleRate * durMs / 1000).round();
    final gapSamples = (_sampleRate * gapMs / 1000).round();
    final total = (beepSamples + gapSamples) * count;
    final data = Int16List(total);

    for (int b = 0; b < count; b++) {
      final offset = b * (beepSamples + gapSamples);
      for (int i = 0; i < beepSamples; i++) {
        final t = i / _sampleRate;
        final env = max(0, 1.0 - i / beepSamples);
        data[offset + i] = (volume * env * sin(2 * pi * freq * t) * 32767).round().clamp(-32768, 32767);
      }
    }
    return _encodeWav(data);
  }

  /// Descending chime of multiple notes
  static Uint8List _chimeWav(List<double> freqs, int durMs, int gapMs, double volume) {
    final noteSamples = (_sampleRate * durMs / 1000).round();
    final gapSamples = (_sampleRate * gapMs / 1000).round();
    final total = (noteSamples + gapSamples) * freqs.length;
    final data = Int16List(total);

    for (int n = 0; n < freqs.length; n++) {
      final offset = n * (noteSamples + gapSamples);
      final freq = freqs[n];
      for (int i = 0; i < noteSamples; i++) {
        final t = i / _sampleRate;
        final env = max(0, 1.0 - i / noteSamples);
        data[offset + i] = (volume * env * sin(2 * pi * freq * t) * 32767).round().clamp(-32768, 32767);
      }
    }
    return _encodeWav(data);
  }

  /// Two rapid notes with a tiny gap between them.
  static Uint8List _twoNoteWav(
    double freq1,
    double freq2,
    int dur1Ms,
    int dur2Ms,
    double volume,
  ) {
    final gapSamples = (_sampleRate * 15 / 1000).round();
    final n1 = (_sampleRate * dur1Ms / 1000).round();
    final n2 = (_sampleRate * dur2Ms / 1000).round();
    final total = n1 + gapSamples + n2;
    final data = Int16List(total);

    int idx = 0;
    for (int i = 0; i < n1; i++, idx++) {
      final t = i / _sampleRate;
      final env = max(0, 1.0 - i / n1 * 0.3);
      data[idx] = (volume * env * sin(2 * pi * freq1 * t) * 32767).round().clamp(-32768, 32767);
    }
    idx += gapSamples;
    for (int i = 0; i < n2; i++, idx++) {
      final t = i / _sampleRate;
      final env = max(0, 1.0 - i / n2);
      data[idx] = (volume * env * sin(2 * pi * freq2 * t) * 32767).round().clamp(-32768, 32767);
    }

    return _encodeWav(data);
  }

  static Uint8List _generateWav({
    required double frequency,
    required int durationMs,
    required double volume,
  }) {
    final numSamples = (_sampleRate * durationMs / 1000).round();
    final data = Int16List(numSamples);

    for (int i = 0; i < numSamples; i++) {
      final t = i / _sampleRate;
      final envelope = max(0, 1.0 - i / numSamples);
      final sample = (volume * envelope * sin(2 * pi * frequency * t) * 32767).round();
      data[i] = sample.clamp(-32768, 32767);
    }
    return _encodeWav(data);
  }

  static Uint8List _generateSweepWav(double startFreq, double endFreq, int durationMs, double volume) {
    final numSamples = (_sampleRate * durationMs / 1000).round();
    final data = Int16List(numSamples);

    for (int i = 0; i < numSamples; i++) {
      final t = i / _sampleRate;
      final envelope = max(0, 1.0 - i / numSamples);
      final freq = startFreq + (endFreq - startFreq) * (i / numSamples);
      final sample = (volume * envelope * sin(2 * pi * freq * t) * 32767).round();
      data[i] = sample.clamp(-32768, 32767);
    }
    return _encodeWav(data);
  }

  static Uint8List _encodeWav(Int16List samples) {
    final byteRate = _sampleRate * _bitsPerSample ~/ 8;
    final dataSize = samples.length * (_bitsPerSample ~/ 8);
    final fileSize = 36 + dataSize;
    final header = ByteData(44);

    header.setUint8(0, 0x52); header.setUint8(1, 0x49);
    header.setUint8(2, 0x46); header.setUint8(3, 0x46);
    header.setUint32(4, fileSize, Endian.little);
    header.setUint8(8, 0x57); header.setUint8(9, 0x41);
    header.setUint8(10, 0x56); header.setUint8(11, 0x45);
    header.setUint8(12, 0x66); header.setUint8(13, 0x6D);
    header.setUint8(14, 0x74); header.setUint8(15, 0x20);
    header.setUint32(16, 16, Endian.little);
    header.setUint16(20, 1, Endian.little);
    header.setUint16(22, 1, Endian.little);
    header.setUint32(24, _sampleRate, Endian.little);
    header.setUint32(28, byteRate, Endian.little);
    header.setUint16(32, _bitsPerSample ~/ 8, Endian.little);
    header.setUint16(34, _bitsPerSample, Endian.little);
    header.setUint8(36, 0x64); header.setUint8(37, 0x61);
    header.setUint8(38, 0x74); header.setUint8(39, 0x61);
    header.setUint32(40, dataSize, Endian.little);

    final wav = Uint8List(44 + dataSize);
    wav.setRange(0, 44, header.buffer.asUint8List());
    wav.setRange(44, 44 + dataSize, samples.buffer.asUint8List());
    return wav;
  }
}

enum DirectionSignalType {
  alert,
  whistle,
  drum,
  dirForward,
  dirBackward,
  dirLeft,
  dirRight,
}

File: wechat-mini-program/src/engines/toneGenerator.ts
import { DirectionSignalType } from '../models/config';

const SAMPLE_RATE = 44100;
const BITS_PER_SAMPLE = 16;

function encodeWav(samples: Int16Array): ArrayBuffer {
  const dataSize = samples.length * (BITS_PER_SAMPLE / 8);
  const fileSize = 36 + dataSize;
  const byteRate = SAMPLE_RATE * BITS_PER_SAMPLE / 8;
  const buffer = new ArrayBuffer(44 + dataSize);
  const view = new DataView(buffer);

  // RIFF header
  view.setUint8(0, 0x52); view.setUint8(1, 0x49); // "RI"
  view.setUint8(2, 0x46); view.setUint8(3, 0x46); // "FF"
  view.setUint32(4, fileSize, true);
  view.setUint8(8, 0x57); view.setUint8(9, 0x41); // "WA"
  view.setUint8(10, 0x56); view.setUint8(11, 0x45); // "VE"
  view.setUint8(12, 0x66); view.setUint8(13, 0x6D); // "fm"
  view.setUint8(14, 0x74); view.setUint8(15, 0x20); // "t "
  view.setUint32(16, 16, true); // chunk size
  view.setUint16(20, 1, true); // PCM
  view.setUint16(22, 1, true); // mono
  view.setUint32(24, SAMPLE_RATE, true);
  view.setUint32(28, byteRate, true);
  view.setUint16(32, BITS_PER_SAMPLE / 8, true);
  view.setUint16(34, BITS_PER_SAMPLE, true);
  view.setUint8(36, 0x64); view.setUint8(37, 0x61); // "da"
  view.setUint8(38, 0x74); view.setUint8(39, 0x61); // "ta"
  view.setUint32(40, dataSize, true);

  const byteData = new Uint8Array(dataSize);
  const sampleView = new DataView(samples.buffer);
  for (let i = 0; i < dataSize; i++) {
    byteData[i] = sampleView.getUint8(i);
  }
  const out = new Uint8Array(buffer);
  out.set(byteData, 44);
  return buffer;
}

function generateWav(freq: number, durMs: number, vol: number): ArrayBuffer {
  const n = Math.floor(SAMPLE_RATE * durMs / 1000);
  const samples = new Int16Array(n);
  for (let i = 0; i < n; i++) {
    const t = i / SAMPLE_RATE;
    const env = Math.max(0, 1 - i / n);
    samples[i] = Math.max(-32768, Math.min(32767, Math.round(vol * env * Math.sin(2 * Math.PI * freq * t) * 32767)));
  }
  return encodeWav(samples);
}

function generateSweepWav(startFreq: number, endFreq: number, durMs: number, vol: number): ArrayBuffer {
  const n = Math.floor(SAMPLE_RATE * durMs / 1000);
  const samples = new Int16Array(n);
  for (let i = 0; i < n; i++) {
    const t = i / SAMPLE_RATE;
    const env = Math.max(0, 1 - i / n);
    const freq = startFreq + (endFreq - startFreq) * (i / n);
    samples[i] = Math.max(-32768, Math.min(32767, Math.round(vol * env * Math.sin(2 * Math.PI * freq * t) * 32767)));
  }
  return encodeWav(samples);
}

function twoNoteWav(f1: number, f2: number, d1: number, d2: number, vol: number): ArrayBuffer {
  const gapSamples = Math.floor(SAMPLE_RATE * 15 / 1000);
  const n1 = Math.floor(SAMPLE_RATE * d1 / 1000);
  const n2 = Math.floor(SAMPLE_RATE * d2 / 1000);
  const total = n1 + gapSamples + n2;
  const samples = new Int16Array(total);
  let idx = 0;
  for (let i = 0; i < n1; i++, idx++) {
    const t = i / SAMPLE_RATE;
    const env = Math.max(0, 1 - i / n1 * 0.3);
    samples[idx] = Math.max(-32768, Math.min(32767, Math.round(vol * env * Math.sin(2 * Math.PI * f1 * t) * 32767)));
  }
  idx += gapSamples;
  for (let i = 0; i < n2; i++, idx++) {
    const t = i / SAMPLE_RATE;
    const env = Math.max(0, 1 - i / n2);
    samples[idx] = Math.max(-32768, Math.min(32767, Math.round(vol * env * Math.sin(2 * Math.PI * f2 * t) * 32767)));
  }
  return encodeWav(samples);
}

function multiBeepWav(freq: number, durMs: number, gapMs: number, count: number, vol: number): ArrayBuffer {
  const beepSamples = Math.floor(SAMPLE_RATE * durMs / 1000);
  const gapSamples = Math.floor(SAMPLE_RATE * gapMs / 1000);
  const total = (beepSamples + gapSamples) * count;
  const samples = new Int16Array(total);
  for (let b = 0; b < count; b++) {
    const offset = b * (beepSamples + gapSamples);
    for (let i = 0; i < beepSamples; i++) {
      const t = i / SAMPLE_RATE;
      const env = Math.max(0, 1 - i / beepSamples);
      samples[offset + i] = Math.max(-32768, Math.min(32767, Math.round(vol * env * Math.sin(2 * Math.PI * freq * t) * 32767)));
    }
  }
  return encodeWav(samples);
}

function chimeWav(freqs: number[], durMs: number, gapMs: number, vol: number): ArrayBuffer {
  const noteSamples = Math.floor(SAMPLE_RATE * durMs / 1000);
  const gapSamples = Math.floor(SAMPLE_RATE * gapMs / 1000);
  const total = (noteSamples + gapSamples) * freqs.length;
  const samples = new Int16Array(total);
  for (let n = 0; n < freqs.length; n++) {
    const offset = n * (noteSamples + gapSamples);
    const freq = freqs[n];
    for (let i = 0; i < noteSamples; i++) {
      const t = i / SAMPLE_RATE;
      const env = Math.max(0, 1 - i / noteSamples);
      samples[offset + i] = Math.max(-32768, Math.min(32767, Math.round(vol * env * Math.sin(2 * Math.PI * freq * t) * 32767)));
    }
  }
  return encodeWav(samples);
}

// Exported functions
export function tickWav(): ArrayBuffer {
  return generateWav(1800, 40, 0.7);
}

export function countIn3Wav(): ArrayBuffer {
  return generateWav(440, 120, 0.7);
}

export function countIn2Wav(): ArrayBuffer {
  return generateWav(660, 120, 0.7);
}

export function countIn1Wav(): ArrayBuffer {
  return generateWav(880, 120, 0.7);
}

export function countInStartWav(): ArrayBuffer {
  return generateWav(1200, 200, 0.9);
}

export function countdownWarningWav(): ArrayBuffer {
  return multiBeepWav(1000, 60, 50, 3, 0.8);
}

export function countdownFinalWav(): ArrayBuffer {
  return multiBeepWav(1200, 40, 30, 5, 0.9);
}

export function endBellWav(): ArrayBuffer {
  return chimeWav([1200, 1000, 800, 600], 200, 40, 0.9);
}

export function signalWav(type: DirectionSignalType): ArrayBuffer {
  switch (type) {
    case DirectionSignalType.alert:
      return generateSweepWav(600, 1200, 150, 0.9);
    case DirectionSignalType.whistle:
      return generateSweepWav(1200, 2000, 120, 0.85);
    case DirectionSignalType.drum:
      return generateWav(120, 100, 1.0);
    case DirectionSignalType.dirForward:
      return twoNoteWav(800, 1300, 45, 55, 0.95);
    case DirectionSignalType.dirBackward:
      return twoNoteWav(1200, 600, 45, 55, 0.95);
    case DirectionSignalType.dirLeft:
      return generateSweepWav(1600, 2600, 75, 0.9);
    case DirectionSignalType.dirRight:
      return generateSweepWav(2600, 1400, 75, 0.9);
  }
}
